File: packages/gizmos-env/lib/Player.ts
import {
  type Energy,
  type AllGizmoLevel,
  ALL_ENERGY_TYPES,
  BuildMethod,
} from './common'
import { find_build_solutions } from './find_build_solutions'
import {
  ConverterGizmo,
  Gizmo,
  is_upgrade_gizmo,
  is_converter_gizmo,
  is_pick_gizmo,
  is_build_gizmo,
  is_file_gizmo,
  GizmoInfo,
  GizmoType,
  UpgradeGizmo,
  PickGizmo,
  BuildGizmo,
  FileGizmo,
} from './Gizmo'
import type { GizmosEnv } from './GizmosEnv'
import { init_energy_num } from './gizmos_utils'

export type BuildSolution = {
  gizmos: ConverterGizmo[]
  energy_num: Record<Energy, number>
}
const BASE_MAX_ENERGY = 5
const BASE_MAX_FILE = 1
const BASE_RESEARCH_NUM = 3

function calc_total_energy_num(energy_num: Record<Energy, number>) {
  return Object.values(energy_num).reduce((acc, curr) => acc + curr, 0)
}

export type PlayerInfo = {
  index: number
  gizmos: GizmoInfo<AllGizmoLevel>[]
  upgrade_gizmos: GizmoInfo[]
  converter_gizmos: GizmoInfo[]
  pick_gizmos: GizmoInfo[]
  build_gizmos: GizmoInfo[]
  file_gizmos: GizmoInfo<AllGizmoLevel>[]
  level3_gizmos: GizmoInfo[]
  filed: GizmoInfo[]
  point_token: number
  max_energy_num: number
  max_file_num: number
  research_num: number
  build_from_filed_cost_reduction: number
  build_from_research_cost_reduction: number
  energy_num: Record<Energy, number>
  total_energy_num: number
  score: number
}

export class Player {
  env: GizmosEnv
  index: number

  gizmos: Set<Gizmo<AllGizmoLevel>>
  point_token: number

  energy_num: Record<Energy, number>
  filed: Set<Gizmo>

  gizmos_by_type: {
    [GizmoType.PICK]: Set<PickGizmo>
    [GizmoType.BUILD]: Set<BuildGizmo>
    [GizmoType.UPGRADE]: Set<UpgradeGizmo>
    [GizmoType.CONVERTER]: Set<ConverterGizmo>
    [GizmoType.FILE]: Set<FileGizmo>
  }
  get upgrade_gizmos() {
    return this.gizmos_by_type[GizmoType.UPGRADE] as Set<UpgradeGizmo>
  }
  get converter_gizmos() {
    return this.gizmos_by_type[GizmoType.CONVERTER] as Set<ConverterGizmo>
  }
  get pick_gizmos() {
    return this.gizmos_by_type[GizmoType.PICK] as Set<PickGizmo>
  }
  get build_gizmos() {
    return this.gizmos_by_type[GizmoType.BUILD] as Set<BuildGizmo>
  }
  get file_gizmos() {
    return this.gizmos_by_type[GizmoType.FILE] as Set<FileGizmo>
  }

  get level3_gizmos() {
    return [...this.gizmos].filter(g => g.level === 3) as Gizmo[]
  }

  max_energy_num: number
  max_file_num: number
  research_num: number
  build_from_filed_cost_reduction: number
  build_from_research_cost_reduction: number

  get total_energy_num() {
    return calc_total_energy_num(this.energy_num)
  }

  get can_add_energy() {
    return this.max_energy_num > this.total_energy_num
  }

  get avail_gizmos() {
    return [...this.gizmos].filter(g => g.active && !g.used)
  }

  get can_file() {
    return this.filed.size < this.max_file_num
  }

  get can_research() {
    return this.research_num > 0
  }

  get score() {
    const total_gizmo_value = [...this.gizmos].reduce(
      (acc, curr) => acc + curr.get_value(this),
      0
    )
    return total_gizmo_value + this.point_token
  }

  calc_cost_reduction(method: BuildMethod) {
    let reduction = 0
    if (method === BuildMethod.FROM_FILED) {
      reduction += this.build_from_filed_cost_reduction
    } else if (method === BuildMethod.FROM_RESEARCH) {
      reduction += this.build_from_research_cost_reduction
    }
    return reduction
  }

  build_solutions(
    gizmo: Gizmo,
    method: BuildMethod,
    cost_energy_num: Record<Energy, number>,
    cost_converter_gizmos: ConverterGizmo[],
    check_only = false
  ): BuildSolution[] {
    if (this.gizmos.has(gizmo)) {
      console.error('[build_solutions] already built')
      return []
    }
    if (
      this.env.check &&
      cost_converter_gizmos.some(
        g => !g.is_satisfied(gizmo) || !this.gizmos.has(g)
      )
    ) {
      console.error('[build_solutions] invalid gizmo used')
      return []
    }
    const cost_reduction = this.calc_cost_reduction(method)
    const solutions = find_build_solutions(
      gizmo.energy_type,
      gizmo.energy_cost - cost_reduction,
      cost_energy_num,
      cost_converter_gizmos,
      check_only
    )
    return solutions.sort(
      (a, b) =>
        calc_total_energy_num(a.energy_num) -
        calc_total_energy_num(b.energy_num)
    )
  }

  add_energy = (energy: Energy) => {
    if (!this.can_add_energy) return
    this.energy_num[energy] += 1
  }

  add_gizmo = (gizmo: Gizmo<AllGizmoLevel>) => {
    this.gizmos.add(gizmo)
    if (is_pick_gizmo(gizmo)) {
      this.gizmos_by_type[GizmoType.PICK].add(gizmo)
    } else if (is_build_gizmo(gizmo)) {
      this.gizmos_by_type[GizmoType.BUILD].add(gizmo)
    } else if (is_upgrade_gizmo(gizmo)) {
      this.gizmos_by_type[GizmoType.UPGRADE].add(gizmo)
      this.max_energy_num += gizmo.max_energy_num
      this.max_file_num += gizmo.max_file_num
      this.research_num += gizmo.research_num
      this.build_from_filed_cost_reduction +=
        gizmo.build_from_filed_cost_reduction
      this.build_from_research_cost_reduction +=
        gizmo.build_from_research_cost_reduction
    } else if (is_converter_gizmo(gizmo)) {
      this.gizmos_by_type[GizmoType.CONVERTER].add(gizmo)
    } else if (is_file_gizmo(gizmo)) {
      this.gizmos_by_type[GizmoType.FILE].add(gizmo)
    }
  }

  pick_from_file(id: number) {
    const gizmo = this.env.gizmo(id)
    if (!this.filed.has(gizmo)) {
      throw new Error("[pick_from_file] not in player's file")
    }
    this.filed.delete(gizmo)
    return gizmo
  }

  pick(energy: Energy) {
    this.add_energy(energy)
    this.pick_gizmos.forEach(g => g.on_pick(energy))
  }

  file(gizmo: Gizmo) {
    if (!this.can_file) {
      throw new Error('file overflow')
    }
    this.filed.add(gizmo)
    this.file_gizmos.forEach(g => g.on_file())
  }

  build(
    gizmo: Gizmo,
    cost_energy_num: Record<Energy, number>,
    cost_converter_gizmos_id: number[],
    method = BuildMethod.DIRECTLY
  ) {
    const converter_gizmos = [...this.converter_gizmos].filter(gizmo =>
      cost_converter_gizmos_id.includes(gizmo.id)
    )
    if (converter_gizmos.length !== cost_converter_gizmos_id.length) {
      throw new Error('unexpected gizmo(s) used')
    }

    if (
      this.env.check &&
      this.build_solutions(
        gizmo,
        method,
        cost_energy_num,
        converter_gizmos,
        true
      ).length <= 0
    ) {
      throw new Error('no build solution')
    }
    converter_gizmos.forEach(g => g.on_convert(gizmo))
    converter_gizmos.forEach(g => g.used_by(this))
    this.drop(cost_energy_num)
    this.build_gizmos.forEach(g =>
      g.on_build(this, gizmo.level, gizmo.energy_type, method)
    )
    this.add_gizmo(gizmo)
  }

  build_from_filed(
    id: number,
    cost_energy_num: Record<Energy, number>,
    cost_converter_gizmos_id: number[]
  ) {
    const gizmo = this.pick_from_file(id)
    this.build(
      gizmo,
      cost_energy_num,
      cost_converter_gizmos_id,
      BuildMethod.FROM_FILED
    )
  }

  build_from_research(
    gizmo: Gizmo,
    cost_energy_num: Record<Energy, number>,
    cost_converter_gizmos_id: number[]
  ) {
    this.build(
      gizmo,
      cost_energy_num,
      cost_converter_gizmos_id,
      BuildMethod.FROM_RESEARCH
    )
  }

  build_for_free(gizmo: Gizmo) {
    this.build_gizmos.forEach(g =>
      g.on_build(this, gizmo.level, gizmo.energy_type, BuildMethod.DIRECTLY)
    )
    this.add_gizmo(gizmo)
  }

  drop(energy_num: Record<Energy, number>) {
    ALL_ENERGY_TYPES.forEach(energy => {
      if (this.energy_num[energy] < energy_num[energy]) {
        throw new Error('not enough energy to drop')
      }
      this.energy_num[energy] -= energy_num[energy]
    })
    this.env.drop_energy_to_pool(energy_num)
    if (this.total_energy_num > this.max_energy_num) {
      throw new Error('energy overflow after drop')
    }
  }

  use_gizmo(id: number) {
    const gizmo = this.env.u_gizmo(id)
    if (!this.gizmos.has(gizmo)) {
      throw new Error("[use_gizmo] not in player's gizmos")
    }
    this.env.state.gizmos[id].used_by(this)
  }

  reset_gizmos() {
    this.gizmos.forEach(g => g.reset())
  }

  get info(): PlayerInfo {
    return {
      index: this.index,
      gizmos: [...this.gizmos].map(g => g.info),
      upgrade_gizmos: [...this.upgrade_gizmos].map(g => g.info),
      converter_gizmos: [...this.converter_gizmos].map(g => g.info),
      pick_gizmos: [...this.pick_gizmos].map(g => g.info),
      build_gizmos: [...this.build_gizmos].map(g => g.info),
      file_gizmos: [...this.file_gizmos].map(g => g.info),
      level3_gizmos: this.level3_gizmos.map(g => g.info),
      filed: [...this.filed].map(g => g.info),
      point_token: this.point_token,
      max_energy_num: this.max_energy_num,
      max_file_num: this.max_file_num,
      research_num: this.research_num,
      build_from_filed_cost_reduction: this.build_from_filed_cost_reduction,
      build_from_research_cost_reduction:
        this.build_from_research_cost_reduction,
      energy_num: this.energy_num,
      total_energy_num: this.total_energy_num,
      score: this.score,
    }
  }

  constructor({
    env,
    index,
    gizmos,
    point_token,
    energy_num,
    filed,
  }: {
    env: GizmosEnv
    index: number
    gizmos?: Gizmo<AllGizmoLevel>[]
    point_token?: number
    energy_num?: Record<Energy, number>
    filed?: Gizmo[]
  }) {
    this.env = env
    this.index = index
    this.max_energy_num = BASE_MAX_ENERGY
    this.max_file_num = BASE_MAX_FILE
    this.research_num = BASE_RESEARCH_NUM
    this.build_from_filed_cost_reduction = 0
    this.build_from_research_cost_reduction = 0
    this.gizmos = new Set()
    this.gizmos_by_type = {
      [GizmoType.PICK]: new Set(),
      [GizmoType.BUILD]: new Set(),
      [GizmoType.UPGRADE]: new Set(),
      [GizmoType.CONVERTER]: new Set(),
      [GizmoType.FILE]: new Set(),
    }
    ;(gizmos ?? []).forEach(this.add_gizmo)
    this.point_token = point_token ?? 0
    this.energy_num = energy_num ?? init_energy_num()
    this.filed = new Set(filed)
  }
}

File: py/Player.py
from __future__ import annotations
from typing import TYPE_CHECKING, Set, TypedDict, List, Dict, Optional
from gizmos_utils import init_energy_num
from common import ALL_ENERGY_TYPES, BuildMethod, Energy
from find_build_solutions import find_build_solutions
from Gizmo import Gizmo, GizmoInfo, is_upgrade_gizmo, is_converter_gizmo, is_pick_gizmo, is_build_gizmo, is_file_gizmo, GizmoType, Gizmo, BuildGizmo, ConverterGizmo, FileGizmo, PickGizmo, UpgradeGizmo

BASE_MAX_ENERGY = 5
BASE_MAX_FILE = 1
BASE_RESEARCH_NUM = 3

if TYPE_CHECKING:
    from GizmosEnv import GizmosEnv


def calc_total_energy_num(energy_num):
    return sum(energy_num.values())


class PlayerInfo(TypedDict):
    index: int
    gizmos: List[Gizmo]
    upgrade_gizmos: List[GizmoInfo]
    converter_gizmos: List[GizmoInfo]
    pick_gizmos: List[GizmoInfo]
    build_gizmos: List[GizmoInfo]
    file_gizmos: List[GizmoInfo]
    level3_gizmos: List[GizmoInfo]
    filed: List[GizmoInfo]
    point_token: int
    max_energy_num: int
    max_file_num: int
    research_num: int
    build_from_filed_cost_reduction: int
    build_from_research_cost_reduction: int
    energy_num: Dict[Energy, int]
    total_energy_num: int
    score: int


class Player:
    def __init__(self,
                 env: GizmosEnv,
                 index: int,
                 gizmos: Optional[List[Gizmo]] = None,
                 point_token: Optional[int] = None,
                 energy_num: Optional[Dict[Energy, int]] = None,
                 filed: Optional[List[Gizmo]] = None):
        self.env = env
        self.index = index
        self.max_energy_num = BASE_MAX_ENERGY
        self.max_file_num = BASE_MAX_FILE
        self.research_num = BASE_RESEARCH_NUM
        self.build_from_filed_cost_reduction = 0
        self.build_from_research_cost_reduction = 0
        self.gizmos: Set[Gizmo] = set()
        self.gizmos_by_type = {
            GizmoType.PICK: set(),
            GizmoType.BUILD: set(),
            GizmoType.UPGRADE: set(),
            GizmoType.CONVERTER: set(),
            GizmoType.FILE: set(),
        }
        for gizmo in gizmos or []:
            self.add_gizmo(gizmo)
        self.point_token = point_token or 0
        self.energy_num = energy_num or init_energy_num()
        self.filed = set([] if filed is None else filed)

    def add_energy(self, energy: Energy):
        if not self.can_add_energy:
            return
        self.energy_num[energy] += 1

    def add_gizmo(self, gizmo: Gizmo):
        self.gizmos.add(gizmo)
        if is_pick_gizmo(gizmo):
            self.gizmos_by_type[GizmoType.PICK].add(gizmo)
        elif is_build_gizmo(gizmo):
            self.gizmos_by_type[GizmoType.BUILD].add(gizmo)
        elif is_upgrade_gizmo(gizmo):
            self.gizmos_by_type[GizmoType.UPGRADE].add(gizmo)
            self.max_energy_num += gizmo.max_energy_num
            self.max_file_num += gizmo.max_file_num
            self.research_num += gizmo.research_num
            self.build_from_filed_cost_reduction += gizmo.build_from_filed_cost_reduction
            self.build_from_research_cost_reduction += gizmo.build_from_research_cost_reduction
        elif is_converter_gizmo(gizmo):
            self.gizmos_by_type[GizmoType.CONVERTER].add(gizmo)
        elif is_file_gizmo(gizmo):
            self.gizmos_by_type[GizmoType.FILE].add(gizmo)

    def pick_from_file(self, id: int):
        gizmo = self.env.gizmo(id)
        if gizmo not in self.filed:
            raise Exception("[pick_from_file] not in player's file")
        self.filed.remove(gizmo)
        return gizmo

    def pick(self, energy: Energy):
        self.add_energy(energy)
        for g in self.pick_gizmos:
            g.on_pick(energy)

    def file(self, gizmo: Gizmo):
        if not self.can_file:
            raise Exception('file overflow')
        self.filed.add(gizmo)
        for g in self.file_gizmos:
            g.on_file()

    def build(self, gizmo: Gizmo, cost_energy_num: Dict[Energy, int], cost_converter_gizmos_id: List[int], method=BuildMethod.DIRECTLY):
        converter_gizmos = [
            g for g in self.converter_gizmos if g.id in cost_converter_gizmos_id]
        if len(converter_gizmos) != len(cost_converter_gizmos_id):
            raise Exception('unexpected gizmo(s) used')

        if self.env.check and len(self.build_solutions(gizmo, method, cost_energy_num, converter_gizmos, True)) <= 0:
            raise Exception('no build solution')
        for g in converter_gizmos:
            g.on_convert(gizmo)
            g.used_by(self)
        self.drop(cost_energy_num)
        for g in self.build_gizmos:
            g.on_build(self, gizmo.level, gizmo.energy_type, method)
        self.add_gizmo(gizmo)

    def build_from_filed(self, id: int, cost_energy_num: Dict[Energy, int], cost_converter_gizmos_id: List[int]):
        gizmo = self.pick_from_file(id)
        self.build(gizmo, cost_energy_num,
                   cost_converter_gizmos_id, BuildMethod.FROM_FILED)

    def build_from_research(self, gizmo: Gizmo, cost_energy_num: Dict[Energy, int], cost_converter_gizmos_id: List[int]):
        self.build(gizmo, cost_energy_num, cost_converter_gizmos_id,
                   BuildMethod.FROM_RESEARCH)

    def build_for_free(self, gizmo: Gizmo):
        for g in self.build_gizmos:
            g.on_build(self, gizmo.level, gizmo.energy_type,
                       BuildMethod.DIRECTLY)
        self.add_gizmo(gizmo)

    def drop(self, energy_num: Dict[Energy, int]):
        for energy in ALL_ENERGY_TYPES:
            if self.energy_num[energy] < energy_num[energy]:
                raise Exception('not enough energy to drop')
            self.energy_num[energy] -= energy_num[energy]
        self.env.drop_energy_to_pool(energy_num)
        if self.total_energy_num > self.max_energy_num:
            raise Exception('energy overflow after drop')

    def use_gizmo(self, id: int):
        gizmo = self.env.u_gizmo(id)
        if gizmo not in self.gizmos:
            raise Exception("[use_gizmo] not in player's gizmos")
        self.env.state['gizmos'][id].used_by(self)

    def reset_gizmos(self):
        for g in self.gizmos:
            g.reset()

    @property
    def info(self) -> PlayerInfo:
        return {
            'index': self.index,
            'gizmos': [g.info for g in self.gizmos],
            'upgrade_gizmos': [g.info for g in self.upgrade_gizmos],
            'converter_gizmos': [g.info for g in self.converter_gizmos],
            'pick_gizmos': [g.info for g in self.pick_gizmos],
            'build_gizmos': [g.info for g in self.build_gizmos],
            'file_gizmos': [g.info for g in self.file_gizmos],
            'level3_gizmos': [g.info for g in self.level3_gizmos],
            'filed': [g.info for g in self.filed],
            'point_token': self.point_token,
            'max_energy_num': self.max_energy_num,
            'max_file_num': self.max_file_num,
            'research_num': self.research_num,
            'build_from_filed_cost_reduction': self.build_from_filed_cost_reduction,
            'build_from_research_cost_reduction': self.build_from_research_cost_reduction,
            'energy_num': self.energy_num,
            'total_energy_num': self.total_energy_num,
            'score': self.score,
        }

    @property
    def can_add_energy(self):
        return self.max_energy_num > self.total_energy_num

    @property
    def avail_gizmos(self):
        return [g for g in self.gizmos if g.active and not g.used]

    @property
    def can_file(self):
        return len(self.filed) < self.max_file_num

    @property
    def can_research(self):
        return self.research_num > 0

    @property
    def score(self):
        total_gizmo_value = sum(g.get_value(self) for g in self.gizmos)
        return total_gizmo_value + self.point_token

    @property
    def total_energy_num(self):
        return calc_total_energy_num(self.energy_num)

    @property
    def upgrade_gizmos(self) -> Set[UpgradeGizmo]:
        return self.gizmos_by_type[GizmoType.UPGRADE]

    @property
    def converter_gizmos(self) -> Set[ConverterGizmo]:
        return self.gizmos_by_type[GizmoType.CONVERTER]

    @property
    def pick_gizmos(self) -> Set[PickGizmo]:
        return self.gizmos_by_type[GizmoType.PICK]

    @property
    def build_gizmos(self) -> Set[BuildGizmo]:
        return self.gizmos_by_type[GizmoType.BUILD]

    @property
    def file_gizmos(self) -> Set[FileGizmo]:
        return self.gizmos_by_type[GizmoType.FILE]

    @property
    def level3_gizmos(self):
        return list(filter(lambda g: g.level == 3, self.gizmos))

    def calc_cost_reduction(self, method: BuildMethod):
        reduction = 0
        if method == BuildMethod.FROM_FILED:
            reduction += self.build_from_filed_cost_reduction
        elif method == BuildMethod.FROM_RESEARCH:
            reduction += self.build_from_research_cost_reduction
        return reduction

    def build_solutions(self, gizmo: Gizmo, method: BuildMethod, cost_energy_num: Dict[Energy, int], cost_converter_gizmos: List[Gizmo], check_only=False):
        if gizmo in self.gizmos:
            print('[build_solutions] already built')
            return []
        if self.env.check and any(not g.is_satisfied(gizmo) or g not in self.gizmos for g in cost_converter_gizmos):
            print('[build_solutions] invalid gizmo used')
            return []
        cost_reduction = self.calc_cost_reduction(method)
        solutions = find_build_solutions(
            gizmo.energy_type, gizmo.energy_cost - cost_reduction, cost_energy_num, cost_converter_gizmos, check_only)
        return sorted(solutions, key=lambda s: calc_total_energy_num(s['energy_num']))
